zng/record.go

```diff
@@ -32,68 +56,15 @@ func (t *TypeRecord) UnmarshalJSON(data []byte) error {
 	if err := json.Unmarshal(data, &t.Columns); err != nil {
 		return err
 	}
-	t.Key = recordString(t.Columns)
+	Typify(t.Context, t.Columns)
 	return nil
 }
 
-func parseColumn(in string) (string, Column, error) {
-	in = strings.TrimSpace(in)
-	colon := strings.IndexByte(in, byte(':'))
-	if colon < 0 {
-		return "", Column{}, ErrTypeSyntax
-	}
-	//XXX should check if name is valid
-	name := strings.TrimSpace(in[:colon])
-	rest, typ, err := parseType(in[colon+1:])
-	if err != nil {
-		return "", Column{}, err
-	}
-	if typ == nil {
-		return "", Column{}, ErrTypeSyntax
-	}
-	return rest, Column{Name: name, Type: typ}, nil
-}
-
-func match(in, pattern string) (string, bool) {
-	in = strings.TrimSpace(in)
-	if strings.HasPrefix(in, pattern) {
-		return in[len(pattern):], true
-	}
-	return in, false
-}
-
-// parseRecordTypeBody parses a list of record columns of the form "[field:type,...]".
-func parseRecordTypeBody(in string) (string, Type, error) {
-	in, ok := match(in, "[")
-	if !ok {
-		return "", nil, ErrTypeSyntax
-	}
-	var columns []Column
-	for {
-		// at top of loop, we have to have a field def either because
-		// this is the first def or we found a comma and are expecting
-		// another one.
-		rest, col, err := parseColumn(in)
-		if err != nil {
-			return "", nil, err
-		}
-		for _, c := range columns {
-			if col.Name == c.Name {
-				return "", nil, ErrDuplicateFields
-			}
-		}
-		columns = append(columns, col)
-		rest, ok = match(rest, ",")
-		if ok {
-			in = rest
-			continue
-		}
-		rest, ok = match(rest, "]")
-		if ok {
-			return rest, LookupTypeRecord(columns), nil
-		}
-		return "", nil, ErrTypeSyntax
-	}
+// FinishUnmarsal is called after the Column types have been filled in
+// by the type resolver.
+func (t *TypeRecord) FinishUnmarsal(data []byte) {
```

Comment:
```suggestion
func (t *TypeRecord) FinishUnmarshal(data []byte) {
```

Reply:
Also doesn't look like this function is called anywhere.

Reply:
deleted